### PR TITLE
[ONNX] Fix `argmin` and `argmax` test cases

### PR DESCRIPTION
**Pull request by qqaatw**
Part of #79263

The `keepdim` argument is theoretically ignored when `dim` is not specified (See [docs](https://pytorch.org/docs/stable/generated/torch.argmin.html)). 

Unfortunately the PyTorch implementation seems to still take it into account, resulting in a non-fully-reduced tensor, which is an undefined behavior. Thus, I add `dim` argument to the tests to make the outputs between PyTorch and ONNX runtime consistent.

### CODE REVIEW

**Review by justinchuby — APPROVED** [review]
LGTM

**Comment by AllenTiTaiWang** [comment]
raise an issue for the doc in #79953.

**Review by titaiwangms — APPROVED** [review]
💯

**Review by BowenBao — DISMISSED** [review]

**Review by justinchuby — CHANGES_REQUESTED** [review]
Thanks for the link to the doc. So the behavior is defined and expected. @qqaatw could you retain the lines and add your changes as additional tests instead?

**Comment by qqaatw** [comment]
> Thanks for the link to the doc. So the behavior is defined and expected. @qqaatw could you retain the lines and add your changes as additional tests instead?

The `ArgMin` and `Argmax` don't allow `axis=None` AFAIK, so in this case we have to flatten the input beforehand then feed it, which underlyingly loses dimension info. After all, retaining the lines would definitely make the proposed check #79263 fail (unless we intentionally resize the output to make it consistent with PyTorch).

**Comment by justinchuby** [comment]
@BowenBao thoughts?

**Comment by AllenTiTaiWang** [comment]
Personally, I feel like depending on what is the optimization goal of onnx opset, if the gold is to align the functionality as much as we can with torch, then maybe Argmin/Argmax should be modified to support this behavior instead and then add the test case, as I don't think changing the code of input/output format in test case would be wise (it definitely affects a lot of other cases.)

**Comment by qqaatw** [comment]
I actually meant by resizing the output in the symbolic functions, not in the test cases. I agree that modifying the functionality of ArgMin and ArgMax on ONNX side would be the best choice.

**Comment by justinchuby** [comment]
ONNX definitions are set for any particular opsets. IMO If resizing is needed to match the behavior with PyTorch then we should do that

**Comment by qqaatw** [comment]
All should be good now.

File: torch/onnx/symbolic_helper.py [hunk 1]
```diff
@@ -800,6 +800,35 @@ def _interpolate_get_scales_and_mode(g, input, size, scale_factor, mode, align_c
     return scale_factor, mode
 
 
+def _argmin_argmax_helper(g, input, dim, keepdim, op_name):
+    keepdim = _parse_arg(keepdim, "i")
+
+    def op_wrapper(input, axis_i, keepdims_i):
```
Comment (justinchuby) [inline torch/onnx/symbolic_helper.py:806]: I would avoid nested functions and pull this out

File: torch/onnx/symbolic_helper.py [hunk 2]
```diff
@@ -800,6 +800,35 @@ def _interpolate_get_scales_and_mode(g, input, size, scale_factor, mode, align_c
...
+        flattened = _reshape_helper(
+            g, input, g.op("Constant", value_t=torch.tensor([-1]))
+        )
+        output = op_wrapper(flattened, axis_i=0, keepdims_i=False)
+        if keepdim:
+            input_shape = g.op("Shape", input)
+            new_shape = g.op("Div", input_shape, input_shape)
+            output = g.op("Reshape", output, new_shape)
+        return output
+    else:
```
Comment (justinchuby) [inline torch/onnx/symbolic_helper.py:827]: else can be removed since all previous branches return

File: torch/onnx/symbolic_helper.py [hunk 3]
```diff
@@ -800,6 +800,35 @@ def _interpolate_get_scales_and_mode(g, input, size, scale_factor, mode, align_c
...
+        return g.op(op_name, input, axis_i=axis_i, keepdims_i=keepdims_i)
+
+    if _is_none(dim):
+        flattened = _reshape_helper(
+            g, input, g.op("Constant", value_t=torch.tensor([-1]))
+        )
+        output = op_wrapper(flattened, axis_i=0, keepdims_i=False)
+        if keepdim:
+            input_shape = g.op("Shape", input)
+            new_shape = g.op("Div", input_shape, input_shape)
```
Comment (justinchuby) [inline torch/onnx/symbolic_helper.py:824]: I see that we are creating a `ones(n)` tensor here. Is that accurate? @qqaatw If so do we think there should be a more straightforward representation for it? @BowenBao

File: torch/onnx/symbolic_helper.py [hunk 4]
```diff
@@ -800,6 +800,35 @@ def _interpolate_get_scales_and_mode(g, input, size, scale_factor, mode, align_c
     return scale_factor, mode
 
 
+def _argmin_argmax_helper(g, input, dim, keepdim, op_name):
```
Comment (justinchuby) [inline torch/onnx/symbolic_helper.py:803]: Could you help annotate the type here? I would add `@parse_args(...)` to `argmax` etc. so that this function can take python types like `int`

**Review by justinchuby — COMMENTED** [review]
Thanks for the change! Some minor comments

File: torch/onnx/symbolic_helper.py (hunk 3 above)
Comment (qqaatw) [reply torch/onnx/symbolic_helper.py:824]: It's int64-int64 division so pretty accurate I think. I have scrutinized it but didn't come up any more concise and straightforward approach.

File: torch/onnx/symbolic_helper.py (hunk 4 above)
Comment (qqaatw) [reply torch/onnx/symbolic_helper.py:803]: I'm wondering can `@parse_args` handle `dim=None`?

File: torch/onnx/symbolic_helper.py (hunk 1 above)
Comment (qqaatw) [reply torch/onnx/symbolic_helper.py:806]: Can you elaborate? I think placing it inside this helper would be more concise than pulling it out.

File: torch/onnx/symbolic_helper.py (hunk 3 above)
Comment (qqaatw) [reply torch/onnx/symbolic_helper.py:824]: Do you have a better solution?

Comment (justinchuby) [reply torch/onnx/symbolic_helper.py:824]: Does `ones_like` work?

File: torch/onnx/symbolic_helper.py (hunk 1 above)
Comment (justinchuby) [reply torch/onnx/symbolic_helper.py:806]: Nested functions are harder to test and maintain, especially when they are a closure. We should eventually get rid of all the opset version conditioning. This particular case is ok though I think. (resolved)

File: torch/onnx/symbolic_helper.py (hunk 3 above)
Comment (justinchuby) [reply torch/onnx/symbolic_helper.py:824]: Does ConstantOfShape work?

Comment (qqaatw) [reply torch/onnx/symbolic_helper.py:824]: 1. Where does the shape come from?
2. It is a static approach I think.

File: torch/onnx/symbolic_helper.py (hunk 4 above)
Comment (justinchuby) [reply torch/onnx/symbolic_helper.py:803]: Probably not at this point but we can add support. For now just parse as "v"?

File: torch/onnx/symbolic_helper.py (hunk 3 above)
Comment (justinchuby) [reply torch/onnx/symbolic_helper.py:824]: What I was thinking was something like

```python
input_shape = g.op("Shape", input)
input_shape_shape = g.op("Shape", input_shape)
new_shape = g.op("ConstantOfShape", input_shape_shape, value_t=torch.tensor( 
    [1], dtype=torch.int64))
```

maybe i was missing something. Can you saw more how it is static?

Comment (qqaatw) [reply torch/onnx/symbolic_helper.py:824]: Thanks for providing the snippet! I think it's feasible and able to handle tensors dynamically.

However, in terms of straightforward, I think your approach is more complex than the current one because IMO it's harder to know the intention of creating  an `ones` tensor without giving a comment.

In addition, though I'm not familiar with the graph implementation of ORT, in general sense your approach may result in more resource usage:

```python
input_shape = g.op("Shape", input) # 1st memory alloc op
input_shape_shape = g.op("Shape", input_shape) # 2nd memory alloc op
new_shape = g.op("ConstantOfShape", input_shape_shape, value_t=torch.tensor( 
    [1], dtype=torch.int64)) # 3rd memory alloc op
```

While the current approach:
```
input_shape = g.op("Shape", input)  # 1st memory alloc op
new_shape = g.op("Div", input_shape, input_shape)  # arithmetic op (if the storage not in-place then 2nd memory alloc)
```

Let me know if I'm missing something.

Comment (justinchuby) [reply torch/onnx/symbolic_helper.py:824]: Thanks for sharing the thought process. I will get more opinions and update here

Comment (justinchuby) [reply torch/onnx/symbolic_helper.py:824]: In principle we prioritize clarity in the conversion results, with the assumption that optimization can be done outside of the conversion process. When we see `Div` we expect some values to be _computed_. In this case it is used for getting a bunch of ones of a certain shape, so as to maintain the rank in the result, which is not really an arithmetic problem (and thus may cause confusion).

An additional inline comment is definitely helpful and does no harm.

Comment (qqaatw) [reply torch/onnx/symbolic_helper.py]: ok.

**Review by justinchuby — APPROVED** [review]
LGTM🚀

**Review by BowenBao — APPROVED** [review]
